shared/src/main/java/com/couchbase/lite/LiveQuery.java

```diff
@@ -0,0 +1,181 @@
+     * NOTE: Calling run method will (re)start the live query.
+     */
+    public void run() {
+        observing = true;
+        releaseResultSet();
+        query.getDatabase().addChangeListener(this);
+        update();
+    }
+
+    /**
+     * Stops observing database changes. Calling start() or getRows() will restart it.
+     */
+    public void stop() {
+        observing = false;
+        willUpdate = false; // cancels the delayed update started by -databaseChanged
+        query.getDatabase().removeChangeListener(this);
+        releaseResultSet();
+    }
+
+    public void addChangeListener(LiveQueryChangeListener listener) {
+        queryChangeListener.add(listener);
+    }
+
+    public void removeChangeListener(LiveQueryChangeListener listener) {
+        queryChangeListener.remove(listener);
+    }
+
+    //---------------------------------------------
+    // Implementation of DatabaseChangeListener
+    //---------------------------------------------
+    @Override
+    public void changed(DatabaseChange change) {
+        if (willUpdate)
+            return; // Already a pending update scheduled
+
+        if (!observing)
+            return;
+
+        // Schedule an update, respecting the updateInterval:
+        long updateDelay = lastUpdatedAt + kDefaultLiveQueryUpdateInterval - System.currentTimeMillis();
+        updateDelay = Math.max(0, Math.min(this.kDefaultLiveQueryUpdateInterval, updateDelay));
+        update(updateDelay);
+    }
+    //---------------------------------------------
+    // protected methods
+    //---------------------------------------------
+
+    @Override
+    protected void finalize() throws Throwable {
+        stop();
+        super.finalize();
+    }
+
+    //---------------------------------------------
+    // Package level access
+    //---------------------------------------------
+
+    //---------------------------------------------
+    // Private (in class only)
+    //---------------------------------------------
+
+    /**
+     * @param delay millisecond
+     */
+    private void update(long delay) {
+        if (willUpdate)
+            return; // Already a pending update scheduled
+
+        if (!observing)
+            return;
+
+        willUpdate = true;
+        // create one doc
+        new Handler(Looper.getMainLooper())
+                .postDelayed(new Runnable() {
+                    @Override
+                    public void run() {
+                        update();
+                    }
+                }, delay); // ms
+    }
+
+    private void update() {
+        // TODO: Make this asynchronous (as in 1.x)
+
+        if (!observing)
+            return;
+
+        try {
+            Log.i(TAG, "%s: Querying...", this);
+            ResultSet oldResultSet = resultSet;
+            ResultSet newResultSet;
+            if (oldResultSet == null || !oldResultSet.isValidEnumerator())
```

Comment:
If it closes automatically, it is not correct to check here as it has a side effect.
For example, 
1. User iterates through the ResultSet -> enumerator closes automatically.
2. There is a database change event but the event doesn't cause the query result to be changed.
3. The update() method is called due to step 2. As the enumerator is not valid anymore, the query is re-run and returns the same old result.